### ISSUE TITLE
Prebuilt ZBM for EFI possibly screws up USB keyboard

### ISSUE THREAD

**Issue by Binarus** [zbm-dev/zfsbootmenu#504]
Dear all,

a few days ago I installed the prebuilt ZBM EFI binary (https://github.com/zbm-dev/zfsbootmenu/releases/download/v2.2.1/zfsbootmenu-release-x86_64-v2.2.1-vmlinuz.EFI) on two of my servers. ZBM works flawlessly there with the following exception:

When ZBM starts the Linux system (Debian bullseye), the first thing the booting Debian system spits out is something like "Tearing down USB controller". When the Debian system has booted, it operates normally except that the USB keyboard is completely dead. I do not know whether it's only the keyboard or whether it's the USB controllers on the mainboard that become defunctional; I guess it's the latter, given the error message.

I do not know whether any previous version of the prebuilt ZBM EFI binary has worked correctly, because until now I generated the ZBM binaries myself. This is currently not possible; therefore I have switched to the prebuilt binaries.

To avoid misunderstandings:

When the servers in question start, they can be normally operated via USB keyboard during the various stages of the boot process, until the ZBM boot menu appears. The USB keyboard also works normally in the ZBM menu itself. However, the Linux kernel that is started by ZBM obviously cannot use the USB controller any more, or at least not the keyboard.

I didn't have that problem with the previous versions of ZBM that I had compiled and generated myself. 

Best regards, and thank you very much,

Binarus

**Comment by zdykstra** [comment]
When you say server, does this machine have an IPMI/iDRAC controller? If so, then this is mostly a matter of conflicting priorities.

A number of XHCI chipsets do not properly reinitialize themselves after a kexec, so the prebuilt images automatically unbind the driver from them just prior to kexec.

A smaller number of systems have the opposite behavior, where the USB interface becomes unusable after this is done. 

We've prioritized supporting the former use-case in our prebuilt releases. Upcoming binary releases will have an easier toggle to allow these types of actions to be disabled easily.  There is a way to do that now, but it's a little more involved.

I'd recommend trying to use the known-good uefi stub linked in the other issue to get a working local build. If you still have trouble with that, we can provide a bit of help to mask the XHCI teardown hook.

**Comment by Binarus** [comment]
Thank you very much for your help!

You are absolutely right. The two servers in question are based on Supermicro X10 platforms that contain IPMI controllers. Thank you very much for your explanation. It might be a good idea to look for a BIOS update then.

I have just tried to get a working local build, but still with the stub that is included in Debian. But this time, the stub is newer, because I am now on bookworm. That didn't change anything. Still a boot prompt with some shell, showing dracut complaining about a missing or empty `rootfs=` variable. My gut feeling is that this is not due a broken EFI stub.

However, I'll try once again with the external known-good EFI stub you linked in the other issue and report back.

Best regards,

Binarus

**Comment by Binarus** [comment]
OK, I just tried again.

I re-compiled zfsbootmenu 2.2.1 from scratch, downloaded the known-good EFI stub, added the path to it in `config.yaml`, mounted the EFI partition to `/boot/efi` and ran `generate-zbm`. This led exactly to the same situation as before.

I am now convinced that this is due to a `dracut` problem, although I didn't touch the `dracut` configuration in any way (compared to that that worked correctly with `generate-zbm` some time ago) and although `generate-zbm -d` didn't show errors that could cause the problem (IMHO).

Could you please give a short outline of the procedure we had to follow if we would like to solve the USB problem? If it involves containers, it would be beyond my horizon. I have experience with VMs and the like, but I have no clue of containers (because I didn't like the underlying concept from the beginning on :-)).

Thank you very much, and best regards,

Binarus

**Comment by ahesford** [comment]
If you're already creating your own images, just follow [these instructions to use a known-good stub loader](https://github.com/zbm-dev/zfsbootmenu/issues/503#issuecomment-1771632259).

If you want to use a prebuilt image, [set `zbm.hookroot`](https://docs.zfsbootmenu.org/en/v2.2.x/man/zfsbootmenu.7.html#command-line-parameters) and mask the `xhci-teardown.sh` script.

**Comment by Binarus** [comment]
At first, thanks you very much!

> If you're already creating your own images, just follow [these instructions to use a known-good stub loader](https://github.com/zbm-dev/zfsbootmenu/issues/503#issuecomment-1771632259).

Well, that's what I have described one post above yours :-) It didn't work even with the known-good stub. When booting, I ran into a `dracut` shell where `dracut` complained that `rootfs=` was empty or missing.

> If you want to use a prebuilt image, [set `zbm.hookroot`](https://docs.zfsbootmenu.org/en/v2.2.x/man/zfsbootmenu.7.html#command-line-parameters) and mask the `xhci-teardown.sh` script.

Actually, I'd like to use my own images. But since I am no longer able to build working images, using prebuilt images is my last resort. I'll try to understand your instructions and to implement them, and of course I'll report back then.

Best regards,

Binarus

**Comment by ahesford** [comment]
Sorry, I overlooked the fact that you already tried the known-good stub loader.

It's not clear to me whether the dracut emergency shell is appearing before ZFSBootMenu loads or after you've selected an environment from within the menu and attempted to boot.

**Comment by Binarus** [comment]
No problem :-)

ZFSBootMenu does *not* appear at all with images I've built. It immediately goes into the `dracut` shell after executing the BIOS. And in that shell, I can't do anything because the keyboard is dead. However, please see my next post :-)

Best regards, and thanks again,

Binarus

**Comment by Binarus** [comment]
> If you want to use a prebuilt image, [set `zbm.hookroot`](https://docs.zfsbootmenu.org/en/v2.2.x/man/zfsbootmenu.7.html#command-line-parameters) and mask the `xhci-teardown.sh` script.

This is ingenious!

I have tried it, and finally it works. I can now boot the prebuilt image and have the keyboard available after Debian has booted. Absolutely great - thanks a lot for your help again.

I guess we now can close this issue. I have tested with only one server so far, but I am sure that it will work for the others as well.
I will keep monitoring the other issue. Perhaps somebody explains there why the images I build myself are failing.

Best regards,

Binarus
